### PR TITLE
Lesson 04 Pull Request

### PR DESCRIPTION
**Pull request by IanMayther**
This is the pull request for the update to the Mailroom to use dictionaries and the trigrams exercise.

### CODE REVIEW

File: students/IanMayther/mailroom/Mailroom.py [hunk 1]
```diff
@@ -62,68 +96,68 @@ def email(TO, gift_amount):
...
+            "3 - Send Letters to ALL Donor.\n"
+            "4 - Quit.\n")
+
+main_selections = {"1" : receiver,
+                    "2" : print_report,
+                    "3" : send_letter,
+                    "4" : quit,
+                    }
+    
+
```
Comment (charlesrobison) [inline students/IanMayther/mailroom/Mailroom.py:160]: Looks good but a couple of functional issues.  Unable to 'list' all donors and when selecting an existing donor, unable to input a donation amount.  Please take another look and let me know if you need help.

File: students/IanMayther/Lesson04/trigrams_exercise.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+
+#Main Exicutable
+if __name__ == '__main__':
+    filename = input("What is the text you would like to input: ")
+    #Please use 'sherlock_small.txt' 
+    
+    in_txt = get_text(filename)
+    trigrams = build_trigrams(in_txt)
+    print(write_text(trigrams))
```
Comment (charlesrobison) [inline students/IanMayther/Lesson04/trigrams_exercise.py:80]: Works well!

**Review by charlesrobison — APPROVED** [review]
A couple of suggestions for mailroom.  Nice job on trigrams.  Note - there are other student's files in this PR, so you'll want to sync with master.

NOT ready to merge - needs to be synced

**Comment by natasha-aleksandrova** [comment]
@IanMayther you have committed merge conflict files, please see files changed tab.

please reach out if you need help resolving this.

**Comment by IanMayther** [comment]
Good Evening Charles,

When you say needs to be sync, do I do that following the git work flow?
Checkout master, pull, push?

Thanks,
Ian

**Comment by natasha-aleksandrova** [comment]
@IanMayther 

https://docs.github.com/en/enterprise/2.15/user/articles/syncing-a-fork

you can use above guide on the commands for syncing with upstream.

**Comment by IanMayther** [comment]
Good Morning Natasha,

I used the link above, but I encountered some issues.  For the branch on
step 2 of the process, I was using
/c/users/ian/SP_Online_PY210/students/IanMayther.  I don't know if that is
too deep?

Then when I; git fetch upstream.  I don't get anything under it, it goes
immediately to the command line.

Finally, when I get down to merging (step 5), I get the error below.

[image: image.png]

I feel like I am one minor step away from resolving this issue.  Please let
me know if you see anything I have done wrong.

Thanks,
Ian


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by natasha-aleksandrova** [comment]
@IanMayther I cannot see images since you've replied from the email - you can comment directly on here on github.com


The commands to sync is something you should be doing as you start work and before you run git add && git commit.

Your current state of fork shows that you've already committed conflicted files after a sync - I did receive your invitation to me as a contributor and I will need to rollback some of your changes that put you into this state. I will be working on that today.

**Comment by IanMayther** [comment]
@natasha-aleksandrova Would it help if I move all of the files I have worked on Lesson05 and Lesson06 and you reset everything back to a previous commit?

**Comment by natasha-aleksandrova** [comment]
@IanMayther let's have you back them up just in case.

**Comment by IanMayther** [comment]
Ian,

I just finished fixing your master branch.

IMPORTANT - do not make any commits, first you will want to make sure that you pull changes down to your clone or re-do your clone.

git pull origin master
[ if you get an error about conflict ]
git checkout --theirs .

then: git add && git commit && git push

You can also re-do your clone via git clone then add upstream again.

Make sure that you're syncing with upstream before adding your new work.
https://docs.github.com/en/enterprise/2.15/user/articles/syncing-a-fork


Let me know if you have any further questions,
Natasha.
________________________________

**Comment by IanMayther** [comment]
@natasha-aleksandrova  Thank you for your help.  I had an issue with the --theirs.  I got a message about --theirs and --ours.  I re-cloned my repo following the steps from https://uwpce-pythoncert.github.io/PythonCertDevel/modules/GitWorkflow.html#git-workflow.  I was able to move my files, I haven't confirmed if they are the same or not.  At worst it doesn't look like too much would have been lost.

I will be more diligent about syncing before starting work.

Thanks again,
Ian

**Comment by natasha-aleksandrova** [comment]
that's great to hear!

**Pull request merged**